include/pybind11/type_caster_pyobject_ptr.h

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2023 The pybind Community.
+
+#pragma once
+
+#include "detail/common.h"
+#include "detail/descr.h"
+#include "cast.h"
+#include "pytypes.h"
+
+PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
+PYBIND11_NAMESPACE_BEGIN(detail)
+
+template <>
+class type_caster<PyObject> {
+public:
+    static constexpr auto name = const_name("PyObject *");
```

Comment:
Won't this const name break mypy/ type parsing? `*` has special meaning and Python itself cannot return pointers.

Reply:
As unsatisfying as it, I think the name should be an `object` caster to avoid breaking anything.

Reply:
It will likely break any position args etc that come after it, right?

Reply:
> As unsatisfying as it, I think the name should be an `object` caster to avoid breaking anything.

Wow ... that's a tough one to decide.

From a high-level perspective, using `object` here definitely looks much better.

But it hides an important clue: beware, this is dealing with raw pointers.

But then again, to whom is that clue useful? The author of the code with the raw pointers already needs to be super careful when writing the code, i.e. doesn't need any extra clues. For everybody else is probably a not very enlightening surprise.

I'm leaning towards thinking using `object` as the name will be better overall.

Reply:
Testing 5ccb893972b921605472dafeffcce6712e5aee94 now.

Reply:
Actually one typing detail. Does this caster accept Py_None as meaning nullptr? That would be the one caveat and if did, it may better to list the type as `typing.Optional`

Reply:
Do you mean the `load()` step?

That's essentially just doing an `inc_ref` and nothing else:

```
    bool load(handle src, bool) {
        value = reinterpret_borrow<object>(src);
        return true;
    }
```